vm/src/builtins/genericalias.rs

```diff
@@ -0,0 +1,136 @@
+use crate::builtins::PyStr;
+use crate::builtins::PyTuple;
+use crate::builtins::PyTypeRef;
+use crate::common::hash;
+use std::fmt;
+
+use crate::slots::{Hashable, SlotConstructor};
+use crate::{
+    IdProtocol, PyClassImpl, PyContext, PyObjectRef, PyRef, PyResult, PyValue, VirtualMachine,
+};
+
+#[pyclass(module = false, name = "types.GenericAlias")]
+pub struct PyGenericAlias {
+    origin: PyTypeRef,
+    args: PyObjectRef,
+    parameters: PyObjectRef,
+}
+
+pub type PyGenericAliasRef = PyRef<PyGenericAlias>;
+
+impl fmt::Debug for PyGenericAlias {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("GenericAlias")
+    }
+}
+
+impl PyValue for PyGenericAlias {
+    fn class(vm: &VirtualMachine) -> &PyTypeRef {
+        &vm.ctx.types.generic_alias_type
+    }
+}
+
+#[derive(FromArgs)]
+pub struct GenericAliasArgs {
+    #[pyarg(any)]
+    origin: PyTypeRef,
+    #[pyarg(any)]
+    arguments: PyObjectRef,
+}
+
+impl SlotConstructor for PyGenericAlias {
+    type Args = GenericAliasArgs;
+
+    fn py_new(cls: PyTypeRef, args: Self::Args, vm: &VirtualMachine) -> PyResult {
+        PyGenericAlias::new(args.origin, args.arguments, vm).into_pyresult_with_type(vm, cls)
+    }
+}
+
+#[pyimpl(with(Hashable), flags(BASETYPE))]
+impl PyGenericAlias {
+    pub fn new(origin: PyTypeRef, args: PyObjectRef, vm: &VirtualMachine) -> Self {
+        let args = if args.downcast_ref::<PyTuple>().is_some() {
+            args
+        } else {
+            vm.ctx.new_tuple(vec![args])
+        };
+
+        Self {
+            origin,
+            args: args.clone(),
```

Comment:
if this is always downcasted to PyObjectRef, I think its type must be `PyTupleRef` instead of `PyObjectRef`